### PR TITLE
RavenDB-21164: Suggestions will now use the whole set of potential matches

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-21164

### Additional description
The method we were using was doing early pruning of the matches, which caused several different issues.

### Type of change
- Bug fix

### How risky is the change?
- Low 

### Backward compatibility
- Non breaking change

### CODE REVIEW

File: src/Corax/Queries/SuggestionTermProvider.cs [hunk 1]
```diff
@@ -156,232 +151,116 @@ public void Dispose()
...
+            var values = new FastList<(Slice Term, int Popularity)>();
+            var dictionary = new Dictionary<uint, int>();
 
-                int take = provider._take;
-                if (take == Constants.IndexSearcher.TakeAll)
-                    take = int.MaxValue;
-                take = Math.Min(Math.Min(take, values.Count), tokens.Length);
+            // We initialize the iterator and store it in the stack memory.
+            Slice.From(allocator, $"__Suggestion_{_fieldId}", out var treeName);
+            var tree = _searcher.Transaction.CompactTreeFor(treeName);
```
Comment (ayende) [inline src/Corax/Queries/SuggestionTermProvider.cs:176]: Who creates this one?

Comment (redknightlois) [reply src/Corax/Queries/SuggestionTermProvider.cs:176]: It was already there :)

**Comment by arekpalinski** [comment]
Can we have some new tests here?

**Comment by arekpalinski** [comment]
Fast tests are failing: https://github.com/ravendb/ravendb/actions/runs/6243411276/job/16948784929

Slow tests: https://jenkins.hibernatingrhinos.com/job/PR_Tests/8326/

**Comment by redknightlois** [comment]
Yes, I just push it in case you could look at it overnight. By 1AM said, screw it... gonna go to sleep

File: test/SlowTests/Issues/RavenDB_295.cs [hunk 2]
```diff
@@ -19,15 +21,21 @@ public RavenDB_295(ITestOutputHelper output) : base(output)
         {
         }
 
-        [Fact]
-        public void CanUpdateSuggestions()
+        public class User
```
Comment (arekpalinski) [inline test/SlowTests/Issues/RavenDB_295.cs]: Please mark it `private`

File: src/Corax/IndexWriter.cs [hunk 3]
```diff
@@ -1226,21 +1224,27 @@ private void AddSuggestions(IndexedField field, Slice slice)
         {
             _hasSuggestions = true;
             field.Suggestions ??= new Dictionary<Slice, int>();
-            
-            var keys = SuggestionsKeys.Generate(_entriesAllocator, Constants.Suggestions.DefaultNGramSize, slice.AsSpan(), out int keysCount);
-            int keySizes = keys.Length / keysCount;
+
+            Span<int> termsLength = stackalloc int[slice.Size];
```
Comment (ayende) [inline src/Corax/IndexWriter.cs]: This can be _really_ big, mind. We may blow the stack here.

File: test/SlowTests/Issues/RavenDB_295.cs [hunk 4]
```diff
@@ -46,26 +54,32 @@ public void CanUpdateSuggestions()
...
-                //Assert.NotEmpty(suggestionQueryResult.Suggestions);
+                using (var s = store.OpenSession())
+                {
+                    var suggestionQueryResult =
+                        s.Advanced.DocumentQuery<User>("test")
+                            .SuggestUsing(x => x.ByField(y => y.Name, "orne"))
+                            .Execute();
+
+                    Assert.NotEmpty(suggestionQueryResult["Name"].Suggestions);
+                }
```
Comment (arekpalinski) [inline test/SlowTests/Issues/RavenDB_295.cs:82]: Can we have a test like in the isssue. That when searching for "aducation" it suggest "education", "duration" etc. So we can know suggestions starting with different letter can be also returned

File: src/Corax/IndexWriter.cs [hunk 5]
```diff
@@ -1251,21 +1255,27 @@ private void RemoveSuggestions(IndexedField field, ReadOnlySpan<byte> sequence)
             field.Suggestions ??= new Dictionary<Slice, int>();
 
 
-            var keys = SuggestionsKeys.Generate(_entriesAllocator, Constants.Suggestions.DefaultNGramSize, sequence, out int keysCount);
-            int keySizes = keys.Length / keysCount;
+            Span<int> termsLength = stackalloc int[sequence.Length];
```
Comment (ayende) [inline src/Corax/IndexWriter.cs]: Here the caller is protecting us, but let's be safe against very large values gain

File: test/SlowTests/Issues/RavenDB_295.cs (hunk 2 above)
Comment (redknightlois) [reply test/SlowTests/Issues/RavenDB_295.cs]: Will do. Can you look at the second test (which I didn't fix)? Could not make it work. If of no use anymore we can just remove it.

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**